File: js/src/decoder.ts
import { IHandler } from "./parser";

export class Decoder implements IHandler<object> {
  private stack: Array<object>;

  constructor() {
    this.stack = [{}];
  }

  // Add a uint64 to the current object
  public uint64(id: number, value: number) {
    this.currentObject()[id] = value;
  }

  // Add binary data to the current object
  public binary(id: number, value: Uint8Array) {
    this.currentObject()[id] = value;
  }

  // Push down the internal stack, constructing a new object
  public beginNested() {
    this.stack.push({});
  }

  // End a nested object, setting it to the given ID on its parent
  public endNested(id: number) {
    let value = this.stack.pop();

    if (this.stack.length === 0) {
      throw new Error("not inside a nested message");
    }

    this.currentObject()[id] = value;
  }

  // Finish decoding, returning the finished parent object
  public finish(): any {
    let result = this.stack.pop();

    if (this.stack.length !== 0) {
      throw new Error("objects remaining in stack");
    }

    return result;
  }

  private currentObject(): any {
    return this.stack[this.stack.length - 1];
  }
}

File: go/decoder.go
package veriform

type decoder struct {
	stack []*Object
}

func NewDecoder() *decoder {
	return &decoder{
		[]*Object{NewObject()},
	}
}

// Called when a uint64 value with the given field ID is parsed
func (d *decoder) Uint64(fieldID FieldID, value uint64) {
	d.currentObject().Store(fieldID, value)
}

// Called when we've received binary data with the given ID
func (d *decoder) Bytes(fieldID FieldID, data []byte) {
	d.currentObject().Store(fieldID, data)
}

// Indicate we've entered a new nested message
func (d *decoder) BeginNested() {
	d.stack = append(d.stack, NewObject())
}

// Indicate we've reached the end of a nested message with the given ID
func (d *decoder) EndNested(fieldID FieldID) {
	if len(d.stack) == 0 {
		panic("not inside a nested message")
	}

	value := d.stack[len(d.stack)-1]
	d.stack = d.stack[:len(d.stack)-1]

	d.currentObject().Store(fieldID, value)
}

// Return the fully parsed object
func (d *decoder) Finish() interface{} {
	if len(d.stack) == 0 {
		panic("message stack is empty")
	} else if len(d.stack) > 1 {
		panic("messages remaining in stack")
	}

	return d.stack[0]
}

// Retrieve the current object on the stack
func (d *decoder) currentObject() *Object {
	return d.stack[len(d.stack)-1]
}
